core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableValues.java

```diff
@@ -70,6 +74,32 @@ public static EnumerableValues create(RelOptCluster cluster,
     return new EnumerableValues(getCluster(), rowType, tuples, traitSet);
   }
 
+  @Override public RelNode passThrough(final RelTraitSet required) {
+    RelCollation collation = required.getCollation();
+    if (collation == null || collation.isDefault()) {
+      return null;
+    }
+    if (tuples.size() > 1) {
+      Ordering<List<RexLiteral>> ordering = null;
+      for (RelFieldCollation fc : collation.getFieldCollations()) {
+        Ordering<List<RexLiteral>> comparator = RelMdCollation.comparator(fc);
+        if (ordering == null) {
+          ordering = comparator;
+        } else {
+          ordering = ordering.compound(comparator);
+        }
+      }
+      if (!ordering.isOrdered(tuples)) {
+        return null;
```

Comment:
Could you also explain which this check?

Reply:
If the tuples are not sorted, we still have to add a sort on it.

Reply:
@chunweilei Are you explaining to @amaliujia or just pointing out a potential bug? :)

Reply:
> @chunweilei Are you explaining to @amaliujia or just pointing out a potential bug? :)

I am explaining to @amaliujia ~~

Reply:
Thanks! Now I see it better!